src/nimgraphviz/private/graphs.nim

```diff
@@ -0,0 +1,304 @@
+import os
+import osproc
+import streams
+import strformat
+import strutils
+
+import "./edges"
+
+import tables
+export tables
+
+
+type
+  NodeCollection* = ref object of RootObj
+    ## Base type for both types of graphs.
+    ## You should not have to instanciate it directly.
+    name*: string    ## The name of the graph
+    graphAttr*: Table[string, string]  ## A table of key-value pairs
+                       ## describing the layout and
+                       ## appearence of the graph
+    nodeAttrs*: Table[string, Table[string, string]] ## A table of all the nodes and their attributes
+
+  Graph* = ref object of NodeCollection
+    ## This types models a non-oriented graph.
+    subGraphs: seq[Graph] # a graph may have multiple sub graphs
+    edges*: Table[Edge, Table[string, string]]
+
+  DiGraph* = ref object of NodeCollection
+    ## This type models an oriented graph.
+    subGraphs: seq[DiGraph] # a graph may have multiple sub graphs
+    edges*: Table[DiEdge, Table[string, string]]
+
+  GenericGraph* = Graph or DiGraph
+
+
+func newGraph*(): Graph =
+  result = Graph(
+    graphAttr: initTable[string, string](),
+    nodeAttrs: initTable[string, Table[string, string]](),
+    subGraphs: newSeq[Graph](),
+    edges: initTable[Edge, Table[string, string]](),
+  )
+
+func newDiGraph*(): DiGraph =
+  result = DiGraph(
+    graphAttr: initTable[string, string](),
+    nodeAttrs: initTable[string, Table[string, string]](),
+    subGraphs: newSeq[DiGraph](),
+    edges: initTable[DiEdge, Table[string, string]](),
+  )
+
+#TODO: find a way to properly merge the two implementations ? (generics)
```

Comment:
For all these duplicated procedures, just do for example:
```nim
func newSubGraph*[T: GenericGraph](parent: T): T =
  result = newGraph()
  parent.subGraphs.add(result)
```

*However*, and this is precisely why I prefer the variant kind approach, you can *only* add add a subgraph of the same type as the parent, because the two are different implementations. By having a single variant object for all kind of graphs, a parent can simply contain multiple types without dealing with N variadic generics or OOP.

Reply:
Hey,
Thank you for your review !
Indeed, when a graph is oriented, then so must all of its subgraphs be. I actually found out when I tried to have a non-oriented graph to have subgraphs using your implementation. Since your subgraphs were always oriented, `dot` crashed at the image rendering stage. (See [here](http://magjac.com/graphviz-visual-editor/?dot=digraph%20%7B%0A%20%20a-%3Eb%3B%0A%20%20subgraph%20%7B%0A%20%20%20%20%20%20x--y%3B%0A%20%20%7D%0A%7D) for a simple example)

That's why I favored having separate "classes" for the two kinds of graphs, instread of an enum type tag. This way, the "orientedness" of the whole graph is checked at compile-time -- that's one less potential source of error. (Also, you'll never want to change orientedness dynamically, so I'd rather have the compiler deal with that; there's optimization potential on its behalf here, as well.)

Then, once the OOP approach is settled (I concede it's not perfect, but it's the approach I chose), I made a base "class" for all the stuff that hasn't got to do with orientedness - I could have duplicated it in each derivated class, but the point of inheritance is precisely to avoid that, and for good reason.
It's indeed not necessary to use the base class when writing operators; the generic approach always works. However, I figure it's cleaner and more efficient to have virtual methods than generic, because the generic actually gets specialized by the compiler (thus creating duplicates): you end up with two functions doing the same thing, and I'm not sure they get merged later on.
Now, for some (methods and) operators, I was stuck, because the orientedness showed in the signature; thank you for pointing the right way to write these ! It won't work with `newSubGraph` in particular, because the call to `newGraph` or `newDiGraph` depends on the actual argument, but it think can be applied to the other operators that demand it.
I'm not sure how I can show the dependency between Graph/Edge and DiGraph/DiEdge, but I think I saw something about it in the manual - there must be something like a `typeof` operator to do this properly; I'll keep experimenting.

Reply:
Ahh, ok. I wasn't aware of these things. Yes, then using distinct types is a good idea indeed.

Still, I don't see why one should use `ref object of` instead of plain `ref object`.
You still get the same amount of procedures from the compilation here independent of that type.

But well, my main problem is I don't fully see where this is going in terms of additional features etc. So feel free to ignore me here. 

In terms of checking for types, you can use `when` as:

```nim
proc someGeneric[T](x: T): Foo = 
  when T is A:
    # only compiled when T is A
  elif T is B: 
    # ...
```
and so on. Not sure if this is what you meant by using `typeof`.

Reply:
> Still, I don't see why one should use ref object of instead of plain ref object.

I mean to use inheritance to avoid having to manually duplicate the attributes (`name`, `graphAttributes`, etc) in the subclasses; as an added benefit, this allows some methods to become virtual instead of generic. The main point is to avoid duplicating things when it can be avoided. Now if you can have the same effect with another syntx, tell me, I just don't know any other way.

> In terms of checking for types, you can use `when`

I knew about `when`, but not about `is`. That's fine for the body of a function, but what I meant was about the declaration :
(The problem when I have an edge as a parameter of a proc is that the type of one parameter determines the type of the other.)
```nim
proc whateverGraph(gph:Graph, edg:Edge) =
  discard

proc whateverDiGraph(gph:DiGraph, edg:DiEdge) =
  discard
```
I would have wanted to do someting along the line of :
```nim
proc whatever[T:GenericGraph](gph:T, edg:typeof(T.edges.Key)) =
  # !! not valid Nim
  discard

proc whatever[T:GenericGraph](gph:T, edg:T.edges.Key) =
  # I think _this_ is ok ?
  discard
```

> I don't fully see where this is going in terms of additional features

Yeah, maybe I didn't take the time to explain my goals clearly here, sorry :wink:
What I mean to improve is :
* __Enforce `dot` compliance__ (basically I would like the library to be unable to generate an invalid dot file by definition). I mean to achieve by using compile-time features as much as possible (e.g. avoid using `assert` et al.).
* __Make graph creation / manipulation easier__ (easier to write the Nim code _and_ easier to read that same Nim code). The strategy here is to define operators when they improve readability (example: don't make a `setEdgeAttr` proc, but rather a custom `[]=` operator. This way you can see the graph as a big table binding attributes to each edge. // counter-example: `addEdge`: I can't find an operator to represent unambigously this operation, adding a `!~!` operator would only make the code obscure). I'm still experimenting here, I'm not sure all the operators I defined in this PR are pointed.
* __Implement (almost) all the features of `dot`__: atm this means mainly the implemting the nesting of sbugraphs. If I see other missing features I might add them to the to-do list :wink: